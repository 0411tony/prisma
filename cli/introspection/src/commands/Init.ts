import { Command, arg, format, HelpError } from '@prisma/cli'
import { isError } from 'util'
import fs from 'fs'
import path from 'path'
import chalk from 'chalk'
import { printError } from '../prompt/utils/print'
import { highlightTS, link } from '@prisma/sdk'

export const defaultSchema = `// This is your Prisma schema file
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}
`

export const defaultEnv  = `# Environment variables declared in this file are automatically made available to Prisma.
# See the documentation for more detail: https://pris.ly/d/prisma-schema#using-environment-variables

# Prisma supports the native connection string format for Postgres, MySQL and SQLite.
# See the documentation for all the connection string options: https://github.com/prisma/prisma2/tree/master/docs/core/connectors

DATABASE_URL="postgresql://johndoe:johndoe@localhost:5432/mydb?schema=public"`

export class Init implements Command {
  static new(): Init {
    return new Init()
  }

  // static help template
  private static help = format(`
    Setup Prisma for your existing database
    
    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma2 init
  `)

  async parse(argv: string[]): Promise<any> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
    })

    if (isError(args)) {
      return null
    }

    if (args['--help']) {
      return this.help()
    }

    const outputDirName = args._[0]
    if (outputDirName) {
      throw Error('The init command does not take any argument.')
    }

    const outputDir = process.cwd()
    const prismaFolder = path.join(outputDir, 'prisma')

    if (fs.existsSync(path.join(outputDir, 'schema.prisma'))) {
      console.log(
        printError(`File ${chalk.bold('schema.prisma')} already exists in your project.
        Please try again in a project that is not yet using Prisma.
      `),
      )
      process.exit(1)
    }

    if (fs.existsSync(prismaFolder)) {
      console.log(
        printError(`Folder ${chalk.bold('prisma')} already exists in your project.
        Please try again in a project that is not yet using Prisma.
      `),
      )
      process.exit(1)
    }

    if (fs.existsSync(path.join(prismaFolder, 'schema.prisma'))) {
      console.log(
        printError(`File ${chalk.bold('prisma/schema.prisma')} already exists in your project.
        Please try again in a project that is not yet using Prisma.
      `),
      )
      process.exit(1)
    }

    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir)
    }

    if (!fs.existsSync(prismaFolder)) {
      fs.mkdirSync(prismaFolder)
    }

    fs.writeFileSync(path.join(prismaFolder, 'schema.prisma'), defaultSchema)
    fs.writeFileSync(path.join(prismaFolder, '.env'), defaultEnv)

    return `
âœ” Your Prisma schema was created at ${chalk.green('prisma/schema.prisma')}.
  You can now open it in your favorite editor.

Next steps
1. Set the ${chalk.green('DATABASE_URL')} connection string in the ${chalk.green('prisma/.env')} file.
2. Run ${chalk.green('prisma2 introspect')} to test the connection and obtain your data model.
3. Run ${chalk.green('prisma2 generate')} to generate Prisma Client.

You can then start using Prisma Client in your application:

\`\`\`
${highlightTS(`\
import { PrismaClient } from '@prisma/client'
// or const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient()
`)}
\`\`\`

More information in our documentation:
${link('https://pris.ly/d/getting-started')}
    `
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Init.help}`)
    }
    return Init.help
  }
}
