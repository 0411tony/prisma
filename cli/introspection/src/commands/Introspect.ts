import { Command, format, HelpError, getSchemaPath, arg } from '@prisma/cli'
import chalk from 'chalk'
import path from 'path'
import {
  getConfig,
  IntrospectionEngine,
  getDMMF,
  dmmfToDml,
  uriToCredentials,
  ConfigMetaFormat,
  RustPanic,
  ErrorArea,
} from '@prisma/sdk'
import { formatms } from '../util/formatms'
import fs from 'fs'
import { DataSource } from '@prisma/generator-helper'
import { databaseTypeToConnectorType } from '@prisma/sdk/dist/convertCredentials'
import { printDatasources } from '../prompt/utils/printDatasources'
import stripAnsi from 'strip-ansi'
import Debug from 'debug'
const debug = Debug('Introspect')

/**
 * $ prisma introspect
 */
export class Introspect implements Command {
  public static new(): Introspect {
    return new Introspect()
  }

  // static help template
  private static help = format(`
    Introspect a database and save the result to schema.prisma.

    ${chalk.bold('Usage')}

    With an existing schema.prisma
      ${chalk.bold('prisma2 introspect')}

    Instead of saving the result to the filesystem, you can also print it
      ${chalk.bold('prisma2 introspect --print')}

  `)
  private constructor() {}

  // parse arguments
  public async parse(argv: string[], minimalOutput = false): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--url': String,
      '--print': Boolean,
    })

    const log = (...messages) => {
      if (!args['--print']) {
        console.log(...messages)
      }
    }

    if (args instanceof Error) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    let url: string | undefined = args['--url']
    let schemaPath = await getSchemaPath()
    let config: ConfigMetaFormat | undefined
    if (!url) {
      if (!schemaPath) {
        throw new Error(
          `Either provide ${chalk.greenBright(
            '--url',
          )} or make sure that you are in a folder with a ${chalk.greenBright('schema.prisma')} file.`,
        )
      }

      config = await getConfig({
        datamodelPath: schemaPath,
      })

      const datasource = config.datasources[0]
      if (!datasource) {
        throw new Error(
          `Either provide ${chalk.greenBright('--url')} or add a ${chalk.greenBright.bold(
            'datasource',
          )} in the ${chalk.greenBright(path.relative(process.cwd(), schemaPath))} file.`,
        )
      }
      url = datasource.url.value
    }

    const engine = new IntrospectionEngine({
      cwd: schemaPath ? path.dirname(schemaPath) : undefined,
    })

    const basedOn =
      !args['--url'] && schemaPath
        ? ` based on datasource defined in ${chalk.underline(path.relative(process.cwd(), schemaPath))}`
        : ''
    log(`Introspecting${basedOn} â€¦`)

    const before = Date.now()
    let introspectionSchema = ''
    introspectionSchema = await engine.introspect(url)
    engine.stop()

    if (introspectionSchema.trim() === '') {
      throw new Error(`Empty introspection result for ${chalk.underline(url)}`)
    }

    const connectorType = databaseTypeToConnectorType(uriToCredentials(url).type)

    const datasourceString = printDatasources([
      {
        config: {},
        connectorType,
        name: 'db',
        url,
      },
    ])

    introspectionSchema = datasourceString + '\n' + introspectionSchema

    debug('introspectionSchema:')
    debug(introspectionSchema)

    try {
      const dmmf = await getDMMF({ datamodel: introspectionSchema })

      // add the datasource itself to the schema in case no schema.prisma exists yet
      const datasources: DataSource[] = [
        {
          name: 'db',
          config: {},
          connectorType,
          url: {
            value: url,
            fromEnvVar: null,
          },
        },
      ]
      const schema = await dmmfToDml({
        config: config || {
          datasources,
          generators: [],
        },
        dmmf: dmmf.datamodel,
      })

      log(`Done with introspection in ${chalk.bold(formatms(Date.now() - before))}`)

      if (args['--print']) {
        console.log(schema)
      } else {
        if (schemaPath && fs.existsSync(schemaPath)) {
          const backupPath = path.join(path.dirname(schemaPath), 'schema.backup.prisma')
          fs.renameSync(schemaPath, backupPath)
          log(
            `\nMoved existing ${chalk.underline(path.relative(process.cwd(), schemaPath))} to ${chalk.underline(
              path.relative(process.cwd(), backupPath),
            )}`,
          )
        }
        schemaPath = schemaPath || 'schema.prisma'
        fs.writeFileSync(schemaPath, schema)
        log(`Wrote ${chalk.underline(path.relative(process.cwd(), schemaPath))}`)
      }
    } catch (e) {
      console.error(chalk.bold.red(`\nIntrospection failed:`) + chalk.red(` Introspected schema can't be parsed.`))
      if (introspectionSchema) {
        console.log(chalk.bold(`Introspected Schema:\n`))
        console.log(introspectionSchema + '\n')
      }
      throw new RustPanic(
        stripAnsi(e.message),
        stripAnsi(e.message),
        {},
        ErrorArea.INTROSPECTION_CLI,
        undefined,
        introspectionSchema,
      )
    }

    return ''
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Introspect.help}`)
    }
    return Introspect.help
  }
}
