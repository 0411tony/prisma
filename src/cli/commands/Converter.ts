import { Command, arg, isError, format, Env, HelpError } from '@prisma/cli'
import chalk from 'chalk'
import { LiftEngine } from '../../LiftEngine'
import { DefaultParser, DatabaseType, isdlToDmmfDatamodel } from 'prisma-datamodel'
import { isdlToDatamodel2 } from '../../utils/isdlToDatamodel2'

export class Converter implements Command {
  static new(env: Env): Converter {
    return new Converter(env)
  }
  private constructor(private readonly env: Env) {}

  // parse arguments
  async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
    })

    if (isError(args)) {
      return this.help(args.message)
    } else if (args['--help']) {
      return this.help()
    }

    const datamodel = await this.readStdin()
    const engine = new LiftEngine({ projectDir: process.cwd() })
    const parser = DefaultParser.create(DatabaseType.postgres)
    const isdl = parser.parseFromSchemaString(datamodel)
    return isdlToDatamodel2(isdl, [])
  }

  readStdin(): Promise<string> {
    return new Promise(resolve => {
      let input = ''

      process.stdin.on('data', data => {
        input += data.toString()
      })
      process.stdin.once('end', () => {
        resolve(input)
      })
      process.stdin.setEncoding('utf-8')
      process.stdin.resume()
    })
  }

  // help message
  help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Converter.help}`)
    }
    return Converter.help
  }

  // static help template
  private static help = format(`
    Convert a datamodel 1.1 to datamodel 2

    ${chalk.bold('Usage')}

      prisma2 convert

    ${chalk.bold('Options')}

      -h, --help       Displays this help message

    ${chalk.bold('Examples')}

      ${chalk.dim(`$`)} cat old-datamodel.prisma | prisma2 convert > new-datamodel.prisma
  `)
}
