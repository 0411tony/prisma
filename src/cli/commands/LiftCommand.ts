import { Command, Commands } from '../types'
import { arg, isError, format } from '../utils'
import { unknownCommand, HelpError } from '../Help'
import kleur from 'kleur'
import { LiftWatch } from './LiftWatch'
import { Env } from '../Env'

/**
 * Migrate command
 */
export class LiftCommand implements Command {
  static new(cmds: Commands, env: Env): LiftCommand {
    return new LiftCommand(cmds, env)
  }
  private constructor(
    private readonly cmds: Commands,
    private readonly env: Env,
  ) {}

  async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--watch': Boolean,
      '-w': '--watch',
      '--preview': Boolean,
      '-p': '--preview',
    })
    if (isError(args)) {
      return this.help(args.message)
    }
    // display help for help flag or no subcommand
    if (args._.length === 0 || args['--help']) {
      return this.help()
    }
    // check if we have that subcommand
    const cmd = this.cmds[args._[0]]
    if (cmd) {
      return cmd.parse(args._.slice(1))
    }

    if (args['--watch']) {
      return LiftWatch.new(this.env).parse(argv)
    }

    return unknownCommand(LiftCommand.help, args._[0])
  }

  help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${kleur.bold().red(`!`)} ${error}\n${LiftCommand.help}`,
      )
    }
    return LiftCommand.help
  }

  // static help template
  private static help = format(`
    ${kleur.bold('üèãÔ∏è‚Äç')}  Lift - Migrate your database schema and data safely

    ${kleur.bold('Usage')}

      prisma lift [command] [options]

    ${kleur.bold('Options')}

      -n, --name      Name of the migration
      -p, --preview   Preview the migration changes
      -w, --watch     Watch for datamodel changes

    ${kleur.bold('Commands')}

      docs   Open documentation in the browser
      down   Migrate your database down
      help   Display command-specific help
    create   Setup a new migration
        up   Migrate your database up

    ${kleur.bold('Examples')}

      Migrate up to the latest datamodel
      ${kleur.dim(`$`)} prisma lift

      Create new migration folder
      ${kleur.dim(`$`)} prisma lift create

      Rollback a migration
      ${kleur.dim(`$`)} prisma lift down 1

      Preview the next migration without applying
      ${kleur.dim(`$`)} prisma lift up 1 --preview

      Watch for any changes to the datamodel
      ${kleur.dim(`$`)} prisma lift --watch
  `)
}
