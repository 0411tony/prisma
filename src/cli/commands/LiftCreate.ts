import { Command, arg, isError, format, HelpError, Env } from '@prisma/cli'
import prompt from 'prompts'
import chalk from 'chalk'
import { Lift } from '../../Lift'
import path from 'path'
import { serializeFileMap } from '../../utils/serializeFileMap'
import { printFiles } from '../../utils/printFiles'
import { printMigrationId } from '../../utils/printMigrationId'
import fs from 'fs'
import { promisify } from 'util'

const writeFile = promisify(fs.writeFile)

/**
 * $ prisma migrate new
 */
export class LiftCreate implements Command {
  static new(env: Env): LiftCreate {
    return new LiftCreate(env)
  }
  private constructor(private readonly env: Env) {}

  // parse arguments
  async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--name': String,
      '-n': '--name',
      '--preview': Boolean,
      '-p': '--preview',
    })
    if (isError(args)) {
      return this.help(args.message)
    } else if (args['--help']) {
      return this.help()
    }
    const preview = args['--preview'] || false
    const name = preview ? args['--name'] : await this.name(args['--name'])
    const lift = new Lift(this.env.cwd)

    const migration = await lift.create(name, preview)

    if (!migration) {
      return `Everything up-to-date\n` //TODO: find better wording
    }

    const { files, migrationId, newLockFile } = migration

    if (preview) {
      return `\nRun ${chalk.greenBright(
        'prisma lift create --name MIGRATION_NAME',
      )} to create the migration\n`
    }

    const migrationsDir = path.join(this.env.cwd, 'migrations', migrationId)
    await serializeFileMap(files, migrationsDir)
    const lockFilePath = path.join(this.env.cwd, 'migrations', 'lift.lock')
    await writeFile(lockFilePath, newLockFile)

    return `\nWe just created your migration ${printMigrationId(
      migrationId,
    )} in\n\n${chalk.dim(
      printFiles(`migrations/${migrationId}`, files),
    )}\n\nRun ${chalk.greenBright('prisma lift up')} to apply the migration\n`
  }

  // get the name
  async name(name?: string): Promise<string | undefined> {
    if (name === '') return undefined
    if (name) return name
    let response = await prompt({
      type: 'text',
      name: 'name',
      message: `Name of migration`,
      validate: value => value.length,
    })
    return response.name || undefined
  }

  // help message
  help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${LiftCreate.help}`,
      )
    }
    return LiftCreate.help
  }

  // static help template
  private static help = format(`
    Create a new migration

    ${chalk.bold('Usage')}

      prisma migrate create [options]

    ${chalk.bold('Options')}

      -h, --help     Displays this help message
      -n, --name     Name the migration

    ${chalk.bold('Examples')}

      Create a new migration
      ${chalk.dim(`$`)} prisma migrate create

      Create a new migration by name
      ${chalk.dim(`$`)} prisma migrate create --name "add unique to email"

  `)
}
