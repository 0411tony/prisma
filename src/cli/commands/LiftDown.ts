import { arg, Command, format, HelpError, isError } from '@prisma/cli'
import chalk from 'chalk'
import { DownOptions, Lift } from '../../Lift'
import { ensureDatabaseExists } from '../../utils/ensureDatabaseExists'
import { ExperimentalFlagError } from '../../utils/experimental'

export class LiftDown implements Command {
  public static new(): LiftDown {
    return new LiftDown()
  }

  // static help template
  private static help = format(`
    Migrate your database down to a specific state.

    ${chalk.bold('Usage')}

      ${chalk.dim(`$`)} prisma migrate down [<dec|name|timestamp>] --experimental

    ${chalk.bold('Arguments')}

      [<dec>]   go down by an amount [default: 1]

    ${chalk.bold('Options')}

      -h, --help       Displays this help message
      -p, --preview    Preview the migration changes

    ${chalk.bold('Examples')}

      Preview a migration without migrating
      ${chalk.dim(`$`)} prisma migrate down --preview --experimental

      Rollback a migration
      ${chalk.dim(`$`)} prisma migrate down 1 --experimental

      Go down to a migration by timestamp
      ${chalk.dim(`$`)} prisma migrate down 20190605204907 --experimental

      Go down to a migration by name
      ${chalk.dim(`$`)} prisma migrate down "add first_name field" --experimental
  `)
  private constructor() {}

  // parse arguments
  public async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview': Boolean,
      '-p': '--preview',
      '--experimental': Boolean,
    })

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    if (isError(args)) {
      return this.help(args.message)
    } else if (args['--help']) {
      return this.help()
    }

    const lift = new Lift()

    const options: DownOptions = {}

    // TODO add go down by name and timestamp
    if (args._.length > 0) {
      const thisArg = args._[0]
      const maybeNumber = parseInt(thisArg, 10)

      // in this case it's a migration id
      if (isNaN(maybeNumber) || typeof maybeNumber !== 'number') {
        throw new Error(`Invalid migration step ${maybeNumber}`)
      } else {
        options.n = maybeNumber
      }
    }

    await ensureDatabaseExists('unapply', true)

    const result = await lift.down(options)
    lift.stop()
    return result
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${LiftDown.help}`)
    }
    return LiftDown.help
  }
}
