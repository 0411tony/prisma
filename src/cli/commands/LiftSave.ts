import { arg, Command, format, getSchema, getSchemaDir, HelpError, isError } from '@prisma/cli'
import chalk from 'chalk'
import fs from 'fs'
import path from 'path'
import prompt from 'prompts'
import { promisify } from 'util'
import { Lift } from '../../Lift'
import { ensureDatabaseExists } from '../../utils/ensureDatabaseExists'
import { printFiles } from '../../utils/printFiles'
import { printMigrationId } from '../../utils/printMigrationId'
import { serializeFileMap } from '../../utils/serializeFileMap'
import { ExperimentalFlagError } from '../../utils/experimental'

const writeFile = promisify(fs.writeFile)

/**
 * $ prisma migrate save
 */
export class LiftSave implements Command {
  public static new(): LiftSave {
    return new LiftSave()
  }

  // static help template
  private static help = format(`
    Save a migration

    ${chalk.bold('Usage')}

      ${chalk.dim(`$`)} prisma migrate save [options] --experimental

    ${chalk.bold('Options')}

      -h, --help       Displays this help message
      -n, --name       Name the migration
      -c, --create-db  Create the database in case it doesn't exist
      -p, --preview    Get a preview of which migration would be created next

    ${chalk.bold('Examples')}

      Create a new migration
      ${chalk.dim(`$`)} prisma migrate save --experimental

      Create a new migration by name
      ${chalk.dim(`$`)} prisma migrate save --name "add unique to email" --experimental

  `)
  private constructor() {}

  // parse arguments
  public async parse(argv: string[]): Promise<string | Error> {
    // parse the arguments according to the spec
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--name': String,
      '-n': '--name',
      '--preview': Boolean,
      '-p': '--preview',
      '--create-db': Boolean,
      '-c': '--create-db',
      '--experimental': Boolean,
    })
    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }
    if (isError(args)) {
      return this.help(args.message)
    } else if (args['--help']) {
      return this.help()
    }
    const preview = args['--preview'] || false
    await ensureDatabaseExists('create', true, args['--create-db'])

    const lift = new Lift()

    const migration = await lift.createMigration('DUMMY_NAME')

    if (!migration) {
      lift.stop()
      return `Everything up-to-date\n` // TODO: find better wording
    }

    const name = preview ? args['--name'] : await this.name(args['--name'])

    const { files, newLockFile, migrationId } = await lift.save(migration, name, preview)

    if (migration.warnings && migration.warnings.length > 0) {
      console.log(chalk.bold(`\n\n⚠️  There might be data loss when applying the migration:\n`))
      for (const warning of migration.warnings) {
        console.log(chalk(`  • ${warning.description}`))
      }
      console.log()
    }

    if (preview) {
      lift.stop()
      return `\nRun ${chalk.greenBright('prisma lift save --name MIGRATION_NAME')} to create the migration\n`
    }

    await getSchema() // just to leverage on its error handling
    const schemaDir = (await getSchemaDir())! // TODO: Probably getSchemaDir() should return Promise<string> instead of Promise<string | null>

    const migrationsDir = path.join(schemaDir, 'migrations', migrationId)
    await serializeFileMap(files, migrationsDir)
    const lockFilePath = path.join(schemaDir, 'migrations', 'lift.lock')
    await writeFile(lockFilePath, newLockFile)

    lift.stop()

    return `\nLift just created your migration ${printMigrationId(migrationId)} in\n\n${chalk.dim(
      printFiles(`migrations/${migrationId}`, files),
    )}\n\nRun ${chalk.greenBright('prisma2 migrate up --experimental')} to apply the migration\n`
  }

  // get the name
  public async name(name?: string): Promise<string | undefined> {
    if (name === '') {
      return undefined
    }
    if (name) {
      return name
    }
    const response = await prompt({
      type: 'text',
      name: 'name',
      message: `Name of migration`,
    })
    return response.name || undefined
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${LiftSave.help}`)
    }
    return LiftSave.help
  }
}
