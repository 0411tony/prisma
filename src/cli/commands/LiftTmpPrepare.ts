import { Command, Dictionary, format, GeneratorDefinitionWithPackage, HelpError } from '@prisma/cli'
import chalk from 'chalk'
import Debug from 'debug'
import { Lift } from '../../Lift'
import { createDatabase } from '../../liftEngineCommands'
import { ensureDatabaseExists } from '../../utils/ensureDatabaseExists'
import { occupyPath } from '../../utils/occupyPath'
const debug = Debug('tmp-prepare')

/**
 * $ prisma migrate new
 */
export class LiftTmpPrepare implements Command {
  public static new(providerAliases: Dictionary<string>): LiftTmpPrepare {
    return new LiftTmpPrepare(providerAliases)
  }

  // static help template
  private static help = format(`
    Watch local changes and migrate automatically

    ${chalk.bold('Usage')}

      prisma dev
  `)
  private constructor(private readonly providerAliases: Dictionary<string>) {}

  // parse arguments
  public async parse(argv: string[]): Promise<string | Error> {
    debug('running tmp-prepare')
    await occupyPath(process.cwd())
    debug('occupied path')

    const lift = new Lift()
    debug('initialized lift')
    await ensureDatabaseExists('dev', false, true)

    await lift.up({
      short: true,
      autoApprove: true,
    })

    await lift.watchUp({
      providerAliases: this.providerAliases,
      autoApprove: true,
    })

    lift.stop()

    console.log('Done executing tmp prepare')
    process.exit(0)

    return ''
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${LiftTmpPrepare.help}`)
    }
    return LiftTmpPrepare.help
  }
}
