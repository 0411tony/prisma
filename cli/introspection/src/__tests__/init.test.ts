import { join } from 'path'
import mkdir from 'make-dir'
import { promisify } from 'util'
import fs from 'fs'
import rimraf from 'rimraf'
import { Init, defaultSchema } from '../commands/Init'

const del = promisify(rimraf)
const tmp = join(__dirname, '../../tmp')

beforeEach(async () => {
  await del(tmp)
  await mkdir(tmp)
})

beforeEach(async () => {
  await del(tmp)
})

test('is schema wriiten on disk replace', async () => {
  const init = Init.new()
  const result = await init.parse(['tmp'])

  const schema = fs.readFileSync(join(tmp, 'prisma/schema.prisma'), 'utf-8')

  expect(schema).toMatch(defaultSchema)
  expect(result).toMatchSnapshot(result)
})
