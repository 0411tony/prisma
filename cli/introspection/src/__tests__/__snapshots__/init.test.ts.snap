// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is schema wriiten on disk replace: We created [32mprisma/schema.prisma[39m for you.
      Edit it with your favorite editor to update your database connection so Prisma can connect to it.

      When done, run [32mprisma2 introspect[39m to test the connection and introspect the data model from your existing database.
      Then run [32mprisma2 generate[39m to generate a Prisma Client based on this data model that can be used in your application.

      More information in our documentation:
      http://tbd
 1`] = `
"We created [32mprisma/schema.prisma[39m for you.
      Edit it with your favorite editor to update your database connection so Prisma can connect to it.

      When done, run [32mprisma2 introspect[39m to test the connection and introspect the data model from your existing database.
      Then run [32mprisma2 generate[39m to generate a Prisma Client based on this data model that can be used in your application.

      More information in our documentation:
      http://tbd
"
`;
